### PR TITLE
[XLA:GPU] Force nccl command buffer to update even memory pointers do not change

### PR DESCRIPTION
**Pull request by shawnwang18**
This PR force command buffer to do cuda graph update if it contains nccl collective operators, because NCCL requires all devices to make the collective call even in the cuda-graph capturing mode.

### CODE REVIEW

File: xla/service/gpu/runtime/command_buffer_cmd.h [hunk 1]
```diff
@@ -311,6 +317,13 @@ class CommandBufferCmdSequence {
   bool empty() const { return commands_.empty(); }
   size_t size() const { return commands_.size(); }
 
+  bool force_update() const {
+    return std::any_of(commands_.begin(), commands_.end(),
```
Comment (ezhulenev) [inline xla/service/gpu/runtime/command_buffer_cmd.h:321]: absl::c_any_of

File: xla/service/gpu/runtime/command_buffer_cmd.cc [hunk 2]
```diff
@@ -1357,7 +1357,7 @@ BarrierCmd::BufferUsageVector BarrierCmd::buffers() { return {}; }
 CollectiveCmd::CollectiveCmd(ExecutionStreamId execution_stream_id,
                              ExecutionStreamId async_from_stream_id,
                              NcclApi* nccl_api, NcclCollectiveConfig config)
-    : TracedCommandBufferCmd(execution_stream_id),
+    : CommandBufferCmd(execution_stream_id),
```
Comment (ezhulenev) [inline xla/service/gpu/runtime/command_buffer_cmd.cc:1360]: Can keep using TracedCommandBuffer and simply pass capacity=0?

File: xla/service/gpu/runtime/command_buffer_thunk.cc [hunk 3]
```diff
@@ -99,6 +99,10 @@ bool CommandBufferThunk::ExecutorCommandBuffer::ShouldUpdateCommandBuffer(
     }
   }
 
+  if (commands.force_update()) {
```
Comment (ezhulenev) [inline xla/service/gpu/runtime/command_buffer_thunk.cc:102]: just put `if (commands.force_update()) return true;` at the beginning

File: xla/service/gpu/runtime/command_buffer_thunk.cc [hunk 4]
```diff
@@ -156,16 +160,18 @@ absl::Status CommandBufferThunk::Initialize(const InitializeParams& params) {
   // before execution, because command buffers when instantiated will allocate
   // memory on device and this might lead to deadlocks when we have concurrent
   // NCCL operations in flight.
-  if (cmd_buffer->command_buffer->state() ==
```
Comment (ezhulenev) [inline xla/service/gpu/runtime/command_buffer_thunk.cc:159]: I can't recall details, but I think there is a reason why we do it twice

File: xla/service/gpu/runtime/command_buffer_thunk.cc [hunk 5]
```diff
@@ -214,8 +223,9 @@ absl::Status CommandBufferThunk::ExecuteOnStream(const ExecuteParams& params) {
 
   if (cmd_buffer->ShouldUpdateCommandBuffer(commands_, params)) {
     VLOG(3) << "Update command buffer on device #" << executor->device_ordinal()
-            << " by recoding command buffer cmd sequence" << " after "
-            << cmd_buffer->num_executions << " executions since last update"
+            << " by recoding command buffer cmd sequence"
```
Comment (ezhulenev) [inline xla/service/gpu/runtime/command_buffer_thunk.cc:226]: accidental?

File: xla/service/gpu/runtime/command_buffer_thunk.cc [hunk 6]
```diff
@@ -156,6 +160,10 @@ absl::Status CommandBufferThunk::Initialize(const InitializeParams& params) {
   // before execution, because command buffers when instantiated will allocate
   // memory on device and this might lead to deadlocks when we have concurrent
   // NCCL operations in flight.
+
+  // If command buffer is in `kFinalize` state, command buffer sequence is
```
Comment (ezhulenev) [inline xla/service/gpu/runtime/command_buffer_thunk.cc:164]:
```
//
// If command buffer in any other state we check it is has to be updated, i.e. if captured pointers
// changed or command buffer has commands that require update on each call.
```

File: xla/service/gpu/runtime/command_buffer_thunk.cc (hunk 5 above)
Comment (shawnwang18) [reply xla/service/gpu/runtime/command_buffer_thunk.cc]: Fixed, clang auto formatting.

File: xla/service/gpu/runtime/command_buffer_thunk.cc (hunk 6 above)
Comment (shawnwang18) [reply xla/service/gpu/runtime/command_buffer_thunk.cc]: Fixed

**Review by ezhulenev — APPROVED** [review]

**Pull request closed without merging**